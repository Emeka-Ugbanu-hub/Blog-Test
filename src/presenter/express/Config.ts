
import Service from '../../service/Service';
import { LoggerInstance } from 'winston';

export default interface Config {
  readonly morganLogFormat: string,
  readonly morganDirectory: string,
  readonly service: Service,
  readonly logger: LoggerInstance;
}