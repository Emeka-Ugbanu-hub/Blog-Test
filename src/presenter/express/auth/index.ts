import login from './login';
import register from './register';
import forgetPassword from './forgetPassword';

export {
  login, 
  register,
  forgetPassword
}