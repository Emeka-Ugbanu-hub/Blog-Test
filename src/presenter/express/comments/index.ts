import createComment from './createComment';
import getCommentById from './getCommentById';
export {
  createComment,
  getCommentById
}