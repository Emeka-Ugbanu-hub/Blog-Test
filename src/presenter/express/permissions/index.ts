import createPermission from './createPermission';
import getPermissionById from './getPermissionById';
import getPermissions from './getPermissions';
import updatePermission from './updatePermission';
import deletePermissionById from './deletePermissionById';
export {
  getPermissions,
  getPermissionById,
  createPermission,
  updatePermission,
  deletePermissionById
}