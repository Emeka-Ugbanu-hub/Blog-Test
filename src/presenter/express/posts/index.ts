import getPosts from './getPosts';
import createPost from './createPost';
import getPostById from './getPostById';
import updatePost from './updatePost';
import deletePostById from './deletePostById';
export {
  getPosts,
  createPost,
  getPostById,
  updatePost,
  deletePostById
}