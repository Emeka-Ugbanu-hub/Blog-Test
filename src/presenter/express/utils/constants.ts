const OK_200_HTTP_CODE = 200;
const CREATED_201_HTTP_CODE = 201;
const NO_CONTENT_204_HTTP_CODE = 204;
const BAD_REQUEST_400_HTTP_CODE = 400;
const UNAUTHORISED_401_HTTP_CODE = 401;
const FORBIDDEN_403_HTTP_CODE = 403;
const NOT_FOUND_404_HTTP_CODE = 404;
const CONFLICT_409_HTTP_CODE = 409;
const UNPROCESSABLE_ENTITY_422_HTTP_CODE = 422;
const SERVER_ERROR_500_HTTP_CODE = 500;


export {
  OK_200_HTTP_CODE,
  CREATED_201_HTTP_CODE,
  NO_CONTENT_204_HTTP_CODE,
  BAD_REQUEST_400_HTTP_CODE,
  UNAUTHORISED_401_HTTP_CODE,
  FORBIDDEN_403_HTTP_CODE,
  NOT_FOUND_404_HTTP_CODE,
  CONFLICT_409_HTTP_CODE,
  UNPROCESSABLE_ENTITY_422_HTTP_CODE,
  SERVER_ERROR_500_HTTP_CODE,
}