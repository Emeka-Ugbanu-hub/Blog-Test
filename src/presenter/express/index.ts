import { Router, Request, Response} from 'express';
import Config from './Config';
import enhancedRouter from './enhancedRouter';
import {register, login, forgetPassword,resetPassword} from './auth';
import {getPosts, createPost, getPostById, updatePost, deletePostById} from './posts';
import {getUserById, getUsers, updateUser, createUser, deleteUserById} from './users';
import {createComment, getCommentById, getComments, updateComment} from './comments';

export default (config: Config): Router => {
  const router: Router = enhancedRouter(config);

  router.get('/', (req: Request, res: Response) => {
    res.status(200)
       .json({message: "This is where the awesomeness happen..."});
  });

  router.post('/auth/register', register(config));
  router.post('/auth/login', login(config));
  router.post('/auth/forget-password', forgetPassword(config));
  router.post('/auth/reset-password', resetPassword(config));

  router.post('/users', createUser(config));
  router.delete('/users/:user_id', deleteUserById(config));
  router.get('/users/:user_id', getUserById(config));
  router.get('/users', getUsers(config));
  router.patch('/users/:user_id', updateUser(config));
  
  router.get('/posts', getPosts(config));
  router.get('/posts/:post_id', getPostById(config));
  router.delete('/posts/:post_id', deletePostById(config));
  router.post('/posts', createPost(config));
  router.patch('/posts/:post_id', updatePost(config));

  router.get('/comments', getComments(config));
  router.post('/comments', createComment(config));
  router.get('/comments/:comment_id', getCommentById(config));
  router.patch('/comments/:comment_id', updateComment(config));

  return router;
}