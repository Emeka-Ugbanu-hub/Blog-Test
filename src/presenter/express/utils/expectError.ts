import {UNPROCESSABLE_ENTITY_422_HTTP_CODE} from './constants';

interface Response {
  status: number;
  body: any;
}

export default (response: Response, expectedStatusCode: number = UNPROCESSABLE_ENTITY_422_HTTP_CODE) => {
  expect(response.status).toBe(expectedStatusCode);
  expect(response.body).toMatchSnapshot();
}