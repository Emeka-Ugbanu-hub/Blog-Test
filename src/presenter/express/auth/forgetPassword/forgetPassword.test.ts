import initTestsWithMailServer from '../../utils/initTestsWithMailServer';
import {API_ROUTE_V1, DEFAULT_USER_PERMISSIONS} from '../../../../utils/constants';
import {Response} from 'express';
import {CREATED_201_HTTP_CODE, OK_200_HTTP_CODE, UNPROCESSABLE_ENTITY_422_HTTP_CODE, CONFLICT_409_HTTP_CODE} from '../../utils/constants';
import config from '../../../../config';
import {TEST_VALID_LOGIN_USER, TEST_INVALID_EMAIL,TEST_VALID_PASSWORD, TEST_VALID_EMAIL, TEST_TOO_SHORT_PASSWORD, TEST_DIFFERENT_VALID_PASSWORD} from '../../../../utils/testValues';
import * as R  from 'ramda';
import expectError from '../../utils/expectError';

describe(__filename, () => {

  const { service, request } = initTestsWithMailServer();

  it('should fail to log in user without input', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/forget-password`);
    expectError(response);
  });

  it('should fail to log in user when email is invalid', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/forget-password`)
                                  .send({
                                    email: TEST_INVALID_EMAIL
                                  });
    expectError(response);
  });


  it('should get 200 even when email does not exist (for security reasons)', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/forget-password`)
                                  .send({email:TEST_VALID_EMAIL});
    expect(response.status).toBe(OK_200_HTTP_CODE);
  });

  it('should succesfull create reset_password_token for user and send email', async () => {
    
    const registeredUser = await service.register({
      email: TEST_VALID_EMAIL, 
      password: TEST_VALID_PASSWORD
    });
    
    const response = await request.post(`${API_ROUTE_V1}/auth/forget-password`)
                                  .send({
                                    email: TEST_VALID_EMAIL
                                  });
                                  
    const resetPasswordTokens = await service.getUserResetPasswordTokens({
      userId: registeredUser.user.id
    });
    expect(resetPasswordTokens.length).toBe(1);
    expect(response.status).toBe(OK_200_HTTP_CODE);
    /*
      ----
# create the server -- will start automatically
import smtpmock
mock_server = smtpmock.MockSMTPServer("localhost", 25025)

#send a test message
import smtplib
client = smtplib.SMTP("localhost", 25025)
fromaddr = "test.sender@mydomain.com"
toaddrs = ["test.recipient1@mydomain.com", "test.recipient2@mydomain.com"]
content = "test message content"
msg = "From: %s\r\nTo: %s\r\n\r\n%s" % (fromaddr, ", ".join(toaddrs), content)
client.sendmail(fromaddr, toaddrs, msg)
client.quit()

# verify that the message has been recieved
assert(mock_server.received_message_matching("From: .*\\nTo: .*\\n+.+tent"))

# reset the server to be ready for a new test
mock_server.reset()
assert(mock_server.received_messages_count() == 0)
    */
  });
});