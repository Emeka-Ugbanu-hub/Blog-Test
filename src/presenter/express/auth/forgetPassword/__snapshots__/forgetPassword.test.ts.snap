// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/forgetPassword/forgetPassword.test.ts should fail to log in user when email is invalid 1`] = `
Object {
  "errors": Array [
    "Invalid email in 'user.email'. Received 'undefined'",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/forgetPassword/forgetPassword.test.ts should fail to log in user without input 1`] = `
Object {
  "errors": Array [
    "Missing required value in 'user.email'",
  ],
  "message": "Unprocessable entity",
}
`;
